drivers/entropy/entropy_stm32.c

```diff
@@ -357,6 +357,21 @@ static int entropy_stm32_rng_init(const struct device *dev)
 	__ASSERT_NO_MSG(dev_cfg != NULL);
 
 #if CONFIG_SOC_SERIES_STM32L4X
+
+#if !defined(CONFIG_CLOCK_STM32_PLL_SRC_MSI) && \
+	!defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI) && \
+	!defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE)
+
+	/* If no PLLSRC clock, use other available 48 MHz clock source */
+#if defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI)
+	LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_HSI48);
```

Comment:
HSI and HSI48 are 2 different clocks.
`CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI=y` doesn't mean HSI48 is on.
HSI48 requires dedicated activation.
Besides HSI48 is not available on all L4, so you should additionally check RCC_HSI48_SUPPORT.

This being said, would it be required ?
Since these are 2 different clocks, this is basically the same case as initially enabling a different clock (PLLSAI).

All in all, maybe you just need to treat the MSI case (since indeed only one clock is required) and the alternate case where an additional clock is required (PLLSAI). And fix what is missing to get it work (enable MSI if missing).

Reply:
Thanks for your corrections and suggestions. If I understand correctly, you're suggesting either to use MSI, without any PLL, or to us PLLSAI1. Is this correct? I'm drafting this in latest commit.
Anyway, I would add some considerations and possibly additional fixes to find a common solution.

- I was trying to reuse existing configs rather than introducing new ones, but I was probably wrong: the CLK48SEL bits select the RNG/USB clock that is well different than SYSCLK selection, so we would need a new Kconfig here, do you agree?
- In such case anyway the logic and the code would become more complex and depend also on SYSCLK selection. E.g. selecting MSI for RNG source would imply different code if SYSCLK is MSI or not.
- Perhaps a simpler logic like the one we're using now would be preferable?
- The PLLSAI1 usage in any case needs to be fixed with valid multiplier/divisors for different cases, e.g. if PLLSRC is MSI @ 48 MHz is not working now.

Reply:
Besides the above points, in the past days we found other problems. 
My colleagues found that the solution with MSI only @ 48 MHz (working both as SYSCLK and  CLK48SEL for RNG) was generating continuously interrupts. As soon as RNG was enabled, it started generating interrupts with CEIS=1, preventing the application to run. This was happening sometimes, the conditions are not clarified (depending on HW?), but the initialization is always the same.
Reading the Reference manual (e.g section 24.3.7), our current code and drivers code, we found some points that are still not clear.
- The CED (bit 5 in RNG CR) seems not available in STM32L462, so we're assuming that the condition CED = 0 is always true for our platform.
- The RNG clock frequency should never be too low, as we first initialize the MSI in stm32_clock_control_init() and then we use it in entropy_stm32_rng_init(). The condition "rng_clk > rng_hclk/16" should always be true (48>48/16). Some delay in between would help?
- Even if we don't know why we get the interrupt with this condition, the stm32_rng_isr() seemed not to clear the CEIS bit interrupt flag like the RM recommends. In my last commit I tried to fix the error case handling, we're testing it from now on.
- Probably lower prio fix to do: the random_byte_get() function appears is returning valid read RNG values that are interpreted as errors because negative.
- It's unclear to us the statement in table 115 that rng_clk should be asynchronous to rng_hclk: is this a real requirement? Can this be a cause for this error?
- Because of the above asynchronous requirement, we tried using PLLSAI1 with MSI @ 48 MHZ as source, using different multiplicator and divisor to get 48 MHz output for RNG. It seemed to work for some days (even if it appears to be still synchronous), but today we found that the same interrupt with CEIS=1 error still happens.
- Our SYSCLK configuration with MSI is using the MSIPLLEN (auto calibration with LSE, PLL-mode). This helped a lot in having a precise clock, hoping not generating issues like this one.

Reply:
@giansta, would you mind opening a dedicated PR/Issue for this specific change ?
I know both topics are linked, but from what I understand they could be dealt with separately.
This will help merging faster in the end

Reply:
Do you mean a dedicated PR/Issue for the flags clearing in the ISR?

Reply:
Yes